### PR TITLE
Fix wrapping of article header

### LINKED ISSUE #264: URLs shown incorrectly
I'm on the current git head (03d3b2d) and noticed since some while some URLs are broken. I'm on Archlinux and using newsbeuter with TinyTinyRSS.

For example, with [this feed](http://www.cyanogenmod.org/feed), when showing the article "Introducing the Cyanogen Apps Package" I see:

```
Feed: news - Cyanogenmod
Title: Introducing the Cyanogen Apps Package
Author: bsxtn
more-themes-added-regularly
Date: Wed, 11 Nov 2015 01:20:37 +0100
```

The other articles look fine.

With a [Google Alert](https://www.google.com/alerts) feed, I see something like this for every post:

```
Feed: qutebrowser - Google Alert
Title: The-Compiler/qutebrowser
sg=AFQjCNHMzf_FR-_GPL8L3qKPru-gKBpIkg
Date: Tue, 02 Feb 2016 11:45:44 +0100
```

### PR DESCRIPTION
**Pull request by Duncanla**
Fixes #264 

To solve this without breaking highlighting of the title, etc, this performs line wrapping in the list formatter so that it can take place after highlighting has occurred.

### CODE REVIEW

**Comment by Minoru** [comment]
Oh, cool! But before reviewing and merging this, let me bisect for a bit and ensure we didn't have this functionality before—I have a feeling it was there but then broke.

**Comment by Minoru** [comment]
Tried out your code. Doesn't work when `text-width` is not set—the body of the article ends up strung out into a column, one word per line. Highlighting in the header works though, unlike with [my fix](https://github.com/akrennmair/newsbeuter/commit/cc8759e4070478e93650e6b8bdaafb21536dd6fd).

**Comment by Duncanla** [comment]
Thanks for catching that, I've fixed it in my latest commit.

I also pulled in your latest commits which fixed the issue, although as you mentioned this patch doesn't break highlighting when wrapping. If you'd like to roll with my fix, then we could probably revert those latest commits or leave them since they aren't hurting anything as far as I can tell (other than having unnecessary code in there). It may make sense to clean up some of the wrapping code as it seems to try to do it in multiple places.

Also, it may be useful to switch the body of the article to use the new wrapping code in the list formatter (right now it happens in the HTML renderer). That way, regex highlighting in the article will also work across wrapped lines.

**Comment by Minoru** [comment]
On Mon, May 02, 2016 at 04:21:51PM -0700, Duncanla wrote:
>I also pulled in your latest commits which fixed the issue, although as 
>you mentioned this patch doesn't break highlighting when wrapping. If 
>you'd like to roll with my fix, then we could probably revert those 
>latest commits or leave them since they aren't hurting anything as far 
>as I can tell (other than having unnecessary code in there).

The problem with your approach is that it fixes things at the very last 
moment. I'm not sure it's correct; it seems to me that something higher 
up the stack should've worried about this. But I don't yet have a full 
understanding of all related code, so I'm not ready to reject your PR 
either—it might indeed prove to be the correct approach.

>It may make sense to clean up some of the wrapping code as it seems to 
>try to do it in multiple places.

That's exactly what I'm intending to do. Currently we have some mess of 
a system that I don't understand and thus can't hope to maintain 
properly. Your patches highlighted a few places where things happen, and 
I'm trying to puzzle together the flow of the data through all the bits.

I probably won't be able to do anything today, but I'll keep you posted 
on the changes and how they affect your PR.

>Also, it may be useful to switch the body of the article to use the new 
>wrapping code in the list formatter (right now it happens in the HTML 
>renderer). That way, regex highlighting in the article will also work 
>across wrapped lines.

Well, I see two approaches (and a half).

One is like your PR: make things higher up the stack emit long formatted 
lines and let listformatter fit them onto the screen. The problem here 
is that we have lines that shouldn't be split, like links in the header 
(there is an issue for that, two hundred something something… Can't 
check right now, sorry.) Basically: URLs should be emitted as long lines 
so that terminals have a chance to interpret them. On the other hand, 
your approach lets us keep all the formatting code in one place, which 
is far easier for maintenance.

What I'm thinking of is a system where HTML renderer decides how things 
should be formatted and either:

* formats them himself, in which case listformatter will be just 
  a dumb interface to std::vector<std::string>; or
* annotates them (wraps into objects or pairs them with some flags) so 
  that listformatter can tell apart things that should and shouldn't be 
  split into multiple lines.

Basically you already provide an implementation of the first approach, 
and I'm experimenting with the second (on master branch. Maybe should 
stop that :) We'll see how it goes.

-- 
Regards,
Alexander Batischev

PGP key 356961A20C8BFD03
Fingerprint: CE6C 4307 9348 58E3 FD94  A00F 3569 61A2 0C8B FD03

**Comment by Minoru** [comment]
The problem here is that we have lines that shouldn't be split, like links in the header (there is an issue for that, two hundred something something… Can't check right now, sorry.)

I meant #282.

BTW, it's better to rebase your branch onto master, not merge master into your branch—merge commits make it harder to review the changes *you* made.

**Comment by Duncanla** [comment]
I think both of your approaches make sense and I'm happy to help test either one. In my opinion, line-wrapping is just formatting of the content so it fits best in the list formatter and required the least amount of code changes (I'm lazy).

Just to see how things would look, I've pushed some more changes that consolidate all line-wrapping code into the list formatter and thus removed what was in htmlrenderer. This had the nice side effect of fixing highlighting across the line breaks in the article body. It would probably be easy to wrap links in a flag in htmlrenderer to avoid breaking links. I figured this should help to compare with your approach so you can decide which one you want to run with.

**Comment by Minoru** [comment]
Looking through the code and then reviewing your changes, I noticed that you stopped initializing `w` variable in htmlrenderer, but didn't remove it from the header. Handler for `<hr>` still depends on it, though. Any ideas of how to fix this within your approach?

**Comment by Minoru** [comment]
Just noticed that the comment I submitted via email a few days ago was stripped of quotes. I can see them if I edit the post, though, but I wasn't able to make them render =\

**Comment by Minoru** [comment]
Since htmlrenderer is almost 900 lines already (and is a bit broken—see #281), I've decided to go with annotations approach for my experiment branch. Now if I only had energy to actually implement it… ._.

**Comment by Minoru** [comment]
At the moment, `listformatter` serves two purposes: it is used both for actual lists (like feed list) and bodies of text (articles). But these two use-cases are proving to be quite different: overly long list items are simply truncated, while text bodies ought to be properly wrapped.

What I'm going to do (right now, no kidding!) is create a new class, `textformatter`, that will be just like `listformatter`, but without random access (i.e. no `set_line` method) and sensitive to the wrapping modes of lines (so that we can mark lines with hyperlinks non-wrappable, causing them to be wrapped by terminals and thus being clickable.)

Sorry for using this PR as a journal, but that's the easiest way of keeping track of thoughts. Also, it helps actually thinking stuff through. And, of course, public record makes it easier for older me to figure out what the hell did I think back then :)

**Comment by Minoru** [comment]
@Duncanla, I've just noticed that with this code the links aren't underscored (see #292 for a screenshot, although the issue itself is not related to your PR.)

**Comment by Minoru** [comment]
My version of the code ended up causing a ton of regressions and even an occasional segfault, so I spent last few days resurrecting the test suite. It's now up and running, but the coverage is abysmal.

I'm working on submitting coverage reports to Coveralls or Codecov right now (pros: nice reports, easy to track changes in the metric).

When it's done, I'm going to write a decent set of test cases for `htmlrenderer` and `listformatter` (extending this to `textformatter` in my experimental branch). After that, you'll be able to rebase this PR and see if there's any more regressions. In the meantime, I'll try to fix my experimental branch.

Finally, with both branches up-to-date and passing all tests, we'll decide which approach to take.

**Comment by Minoru** [comment]
Tests are ready. You can rebase your code and fix regressions at laisure—I'm so tired of thinking of HTML rendering that I'm probably going to spend tomorrow on something else.

Thanks for sticking with me :)

**Comment by Duncanla** [comment]
> Tests are ready. You can rebase your code and fix regressions at leisure

Thanks, I'll do this as soon as I catch some free time

> Looking through the code and then reviewing your changes, I noticed that you stopped initializing w   variable in htmlrenderer, but didn't remove it from the header. Handler for \<hr\> still depends on it, though. Any ideas of how to fix this within your approach?


> I've just noticed that with this code the links aren't underscored

I think these two issues are going to be hard to solve with the approach I have been taking. Your way could handle it quite nicely though, by tagging \<hr\>'s and links and letting the textformatter do the rest. The only other reasonable way I can see to handle it would be to move the formatting into htmlrenderer but that seems a bit more clumsy than your tagging approach. I think this makes it pretty convincing that you are doing it the best way.

**Comment by Minoru** [comment]
Ah, don't make updating this PR your priority then—let's see what I end up with. Your `wrap_line` function will make it into the codebase anyway as I just copy-pasted it from your PR ;)

**Comment by Minoru** [comment]
I *finally* pushed my changes >_< Sorry for the delay.

Ended up using your wrapping code. Thanks!
